Doc/library/__main__.rst

```diff
@@ -1,25 +1,102 @@
+or from a :option:`-m` argument, but **not** when it is initialized from an
+import statement.  In any of these situations, the module's ``__name__`` is set
+equal to ``'__main__'``.  The only other context in which Python code is run is
+when it is imported through an import statement. In that case, ``__name__`` is
+set equal to the module's name; usually the name of the file without the
+``.py`` extension.
+
+As a result, a module can discover whether or not it is running in the
+top-level environment by checking its own ``__name__``, which allows a common
+idiom for conditionally executing code when the module is not initialized from
+an import statement::
+
+    if __name__ == '__main__':
+        # Execute when the module is not initialized from an import statement.
+        ...
+
+
+``__main__.py`` in Python Packages
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+If you are not familiar with Python packages, see section :ref:`tut-packages`.
+Most commonly, the ``__main__.py`` file is used to provide a command line
+interface for a package. Consider the following hypothetical package,
+"bandclass":
+
+.. code-block:: text
+
+   bandclass
+     ├── __init__.py
+     ├── __main__.py
+     ├── parent.py
+     └── student.py
+
+``__main__.py`` will be executed when the package itself is invoked
+directly from the command line using the :option:`-m` flag. For example::
+
+    python3 -m bandclass
+
+This command will cause ``__main__.py`` to run. How you utilize this
+mechanism will depend on the nature of the package you are writing, but
+in this hypothetical case, it might make sense to allow the teacher to search
+for students or parents using :mod:`argparse`::
+
+    # bandclass/__main__.py
+
+    import argparse
+    import sys
+
+    from .parent import Parents
+    from .student import Students
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--student',
+                        help="lookup a student and print their information")
+    parser.add_argument('--parent',
+                        help="lookup a parent and print their information")
+
+    args = parser.parse_args()
 
-A module can discover whether or not it is running in the main scope by
-checking its own ``__name__``, which allows a common idiom for conditionally
-executing code in a module when it is run as a script or with ``python
--m`` but not when it is imported::
+    if args.student and student := Students.find(args.student):
+        print(student)
+        sys.exit('Student found')
+    elif args.parent and parent := Parents.find(args.parent):
+        print(parent)
+        sys.exit('Parent found')
+    else:
+        print('Result not found')
+        sys.exit(args.print_help())
 
-   if __name__ == "__main__":
-       # execute only if run as a script
-       main()
+Note that there is no reason to use the ``if __name__ == '__main__'`` statement
+in ``__main__.py`` itself. There is no reason for any other file to import
```

Comment:
I sometimes use `console_scripts=package.__main__:main` in which case the `__name__` will be `"package.__main__"`

Reply:
It does say, "in **most cases** it would be a redundant statement." There are definitely exceptions to that rule, but I think that to start listing them out every edge case would make this document unnecessarily complex.

Also, by using `console_scripts` like that, what functionality are you getting from the distributed package that is different from just `python -m package` running `__main__.py` as normal?

Reply:
Loads! Eg `pytest --cov` and `coverage run -m pytest` change the `sys.path` and break expectations of `tox` users

Reply:
Ok, see 7e495d7bba0d4cd0685fc66bc179a3bd88359532. Of course, the other direction to go here is to remove this paragraph all together and let users make their own decisions about how to use `if __name__ == '__main__` blocks in their `__main__.py` files.

If you have a project off the top of your head that you could link me to that uses this pattern, I would appreciate you linking me to it so that I can take a look and get to know what you're talking about better :)